scalanet/discovery/src/io/iohk/scalanet/discovery/ethereum/v4/DiscoveryNetwork.scala

```diff
@@ -0,0 +1,376 @@
+      messageExpiration: FiniteDuration = 60.seconds,
+      // Timeout for individual requests.
+      requestTimeout: FiniteDuration = 3.seconds,
+      // Timeout for collecting multiple potential Neighbors responses.
+      kademliaTimeout: FiniteDuration = 7.seconds,
+      // Max number of neighbours to expect.
+      kademliaBucketSize: Int = 16
+  )(implicit codec: Codec[Payload], sigalg: SigAlg, clock: Clock[Task]): Task[DiscoveryNetwork[A]] = Task {
+    new DiscoveryNetwork[A] with LazyLogging {
+
+      import DiscoveryRPC.ENRSeq
+      import Payload._
+
+      private val expirationMillis = messageExpiration.toMillis
+
+      private val currentTimeMillis = clock.realTime(MILLISECONDS)
+
+      private val maxNeighborsPerPacket = getMaxNeighborsPerPacket
+
+      private val localNodeAddress = toNodeAddress(peerGroup.processAddress)
+
+      /** Start a fiber that accepts incoming channels and starts a dedicated fiber
+        * to handle every channel separtely, processing their messages one by one.
+        * This is fair: every remote connection can be throttled independently
+        * of each other, as well as based on operation type by the `handler` itself.
+        */
+      override def startHandling(handler: DiscoveryRPC[(PublicKey, A)]): Task[CancelableF[Task]] =
+        for {
+          cancelToken <- Deferred[Task, Unit]
+          _ <- peerGroup
+            .nextServerEvent()
+            .withCancelToken(cancelToken)
+            .toIterant
+            .mapEval {
+              case ChannelCreated(channel, release) =>
+                handleChannel(handler, channel, cancelToken)
+                  .guarantee(release)
+                  .onErrorRecover {
+                    case ex: TimeoutException =>
+                    case NonFatal(ex) =>
+                      logger.error(s"Error handling channel from ${channel.to}: $ex")
+                  }
+                  .startAndForget
+
+              case _ =>
+                Task.unit
+            }
+            .completedL
+            .startAndForget
+          cancelable <- CancelableF[Task](cancelToken.complete(()))
+        } yield cancelable
+
+      private def handleChannel(
+          handler: DiscoveryRPC[(PublicKey, A)],
+          channel: Channel[A, Packet],
+          cancelToken: Deferred[Task, Unit]
+      ): Task[Unit] = {
+        channel
+          .nextMessage()
+          .withCancelToken(cancelToken)
+          .timeout(messageExpiration) // Messages older than this would be ignored anyway.
+          .toIterant
+          .mapEval {
+            case MessageReceived(packet) =>
+              currentTimeMillis.flatMap { timestamp =>
+                Packet.unpack(packet) match {
+                  case Attempt.Successful((payload, remotePublicKey)) =>
+                    payload match {
+                      case _: Payload.Response =>
+                        // Not relevant on the server channel.
+                        Task.unit
```

Comment:
No this isn't an error here, this is how the `StaticUDPPeerGroup` works. Similarly to how [KNetwork only collects KRequests](https://github.com/input-output-hk/scalanet/blob/develop/scalanet/kademlia/src/io/iohk/scalanet/kademlia/KNetwork.scala#L58) here we are in the context of the server channel, so we can ignore messages which don't need answers. With the `DynamicUDPPeerGroup` we could have had an expectation that a client connecting to us will send requests, however in the case of the `StaticUDPPeerGroup`  we actually know that the server channel will receive _all_ incoming messages, not just requests, because it can never tell about an incoming packet what it is, they all arrive from the same socket, so it does the replication to all channels.